### PR TITLE
Add HA Support for the Starboard Operator

### PR DESCRIPTION
**Pull request by xyoxo**
Resolves https://github.com/aquasecurity/starboard/issues/219

### CODE REVIEW

File: deploy/helm/templates/rbac.yaml [hunk 1]
```diff
@@ -89,6 +89,14 @@ rules:
       - watch
       - create
       - update
+  - apiGroups:
+      - coordination.k8s.io
```
Comment (danielpacak) [inline deploy/helm/templates/rbac.yaml:93]: Do we need that permission? I see that by default controller-runtime writes to ConfigMap not Lease. Is it somehow configurable?

File: pkg/operator/etc/config.go [hunk 2]
```diff
@@ -19,6 +19,9 @@ type Config struct {
 	HealthProbeBindAddress        string        `env:"OPERATOR_HEALTH_PROBE_BIND_ADDRESS" envDefault:":9090"`
 	LogDevMode                    bool          `env:"OPERATOR_LOG_DEV_MODE" envDefault:"false"`
 	CISKubernetesBenchmarkEnabled bool          `env:"OPERATOR_CIS_KUBERNETES_BENCHMARK_ENABLED" envDefault:"true"`
+	LeaderElectionEnabled         bool          `env:"OPERATOR_LEADER_ELECTION_ENABLED" envDefault:"false"`
+	LeaderElectionNamespace       string        `env:"OPERATOR_LEADER_ELECTION_NAMESPACE"`
```
Comment (danielpacak) [inline pkg/operator/etc/config.go:23]: Can we use existing OPERATOR_NAMESPACE instead? Is there any scenario where we may want to save lock in the different namespace than the operator itself?

**Review by danielpacak — COMMENTED** [review]
👋 @xyoxo Good job. I tested it on my end and it seems to work as expected. I left a few questions though before we merge this PR.

I also noticed that controller-runtime is trying to create K8s events, but starboard-operator service account does not have such permission, i.e. I see the following errors in the log:

E0324 16:05:31.811972       1 event.go:264] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"starboard-operator.166f52e450b2f92f", GenerateName:"", Namespace:"starboard-operator", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"ConfigMap", Namespace:"starboard-operator", Name:"starboard-operator", UID:"04b1b499-fca9-4bad-a905-c6ff6066a802", APIVersion:"v1", ResourceVersion:"1913", FieldPath:""}, Reason:"LeaderElection", Message:"starboard-operator-b7cffd786-7slnw_843590a9-06bb-4aad-b58f-ca6c85a05b7d became leader", Source:v1.EventSource{Component:"starboard-operator-b7cffd786-7slnw_843590a9-06bb-4aad-b58f-ca6c85a05b7d", Host:""}, FirstTimestamp:v1.Time{Time:time.Time{wall:0xc00ef6b2f03bcb2f, ext:531398242, loc:(*time.Location)(0x2336e20)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0xc00ef6b2f03bcb2f, ext:531398242, loc:(*time.Location)(0x2336e20)}}, Count:1, Type:"Normal", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events is forbidden: User "system:serviceaccount:starboard-operator:starboard-operator" cannot create resource "events" in API group "" in the namespace "starboard-operator"' (will not retry!)

File: pkg/operator/etc/config.go (hunk 2 above)
Comment (danielpacak) [reply pkg/operator/etc/config.go:23]: We should also document any new envs in `docs/operator/configuration.md`

Comment (xyoxo) [reply pkg/operator/etc/config.go:23]: I think we can use the `OPERATOR_NAMESPACE`.

File: deploy/helm/templates/rbac.yaml (hunk 1 above)
Comment (xyoxo) [reply deploy/helm/templates/rbac.yaml:93]: controller-runtime uses `configmapleases` (both configmap and lease) by default. They introduced it to migrate from configmap to lease. I think they will remove configmap support in the future.

Comment (xyoxo) [reply deploy/helm/templates/rbac.yaml:93]: I tried giving events create permission, then a new error shows up. I'm not sure why this is showing no kind registered for lease.

```
E0324 18:05:48.829011       1 event.go:329] Could not construct reference to: '&v1.Lease{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"starboard-operator", GenerateName:"", Namespace:"starboard-operator", SelfLink:"/apis/coordination.k8s.io/v1/namespaces/starboard-operator/leases/starboard-operator", UID:"7b828f2b-235a-4034-911a-d3bc6c1dc231", ResourceVersion:"379025", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63752121432, loc:(*time.Location)(0x22f3080)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"starboard-operator", Operation:"Update", APIVersion:"coordination.k8s.io/v1", Time:(*v1.Time)(0xc000117758), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0xc000117770)}}}, Spec:v1.LeaseSpec{HolderIdentity:(*string)(nil), LeaseDurationSeconds:(*int32)(nil), AcquireTime:(*v1.MicroTime)(nil), RenewTime:(*v1.MicroTime)(nil), LeaseTransitions:(*int32)(nil)}}' due to: 'no kind is registered for the type v1.Lease in scheme "pkg/runtime/scheme.go:101"'. Will not report event: 'Normal' 'LeaderElection' 'starboard-starboard-operator-57cd876b88-2jn2j_5a1380bc-b308-49a8-b64d-fef54a645a32 became leader'
```

Comment (danielpacak) [reply deploy/helm/templates/rbac.yaml:93]: I think we just need to add coordination/v1 GV to `runtime.Scheme` returned by the `starboard.NewScheme()` factory. I tried that and got rid of all errors:

fix

**Review by danielpacak — APPROVED** [review]

**Pull request merged**